### PR TITLE
PLAN-120 SP counting

### PR DESCRIPTION
**Pull request by RutaB98**

### CODE REVIEW

File: E2E/assertions/assertions.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
+const { I, game } = inject();
+const complexitySP = [1,2,3,5,8,13];
```
Comment (jvalatkevicius) [inline E2E/assertions/assertions.ts:2]: using arrays for these kind of values is extremely confusing as `complexitySP[1]` would result to `2`.

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: E2E/pageObjects/game.page.ts [hunk 2]
```diff
@@ -85,12 +89,25 @@ export = {
...
     I.click(locator.buttons.skip);
     I.seeElement(locator.buttons.skipButtonHighlighted);
   },
+
   voteAndCheckThatVoteIsVisible:(chessPiece: ChessPiece, tile: string, value:string) => {
     game.vote(chessPiece, tile);
     game.voteIsVisible(chessPiece, tile, value);
   },
+
+  checkPlayerSP:(averageScore) => {
```
Comment (AndzejIlj) [inline E2E/pageObjects/game.page.ts:103]: Check what happens when player's average score is 1,33. It should return 1, yet your method returns 0

**Review by AndzejIlj — CHANGES_REQUESTED** [review]

File: E2E/pageObjects/game.page.ts [hunk 3]
```diff
@@ -98,12 +102,31 @@ export = {
...
```
Comment (AndzejIlj) [inline E2E/pageObjects/game.page.ts:116]: When rounding up the average score, if delta is the same to lower and higher fibonacci number, it should return the higher number, your method does the opposite. For example, if `averageScore` is 4, it should return 5, but test returns 3.

**Review by AndzejIlj — CHANGES_REQUESTED** [review]

**Review by KristinaKrev — APPROVED** [review]

File: E2E/assertions/assertions.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+export = {
+    async checkIfTheLinkWasCopied (currentUrl) {
+      I.usePlaywrightTo('check clipboard', async ({ page, browserContext }) => {
+        await browserContext.grantPermissions(['clipboard-read']);
+        const clipboardText = await page.evaluate("navigator.clipboard.readText()");
+        I.assertEqual(currentUrl, clipboardText);
+      });
+  },
+
+  async checkIfTheSPIsRoundedCorrectlyForTheFirstPlayer() {
```
Comment (jvalatkevicius) [inline E2E/assertions/assertions.ts:12]: why is this extracted? If I understand correctly, this method is only valid when some specific conditions are met (e.g. placing some predefined figures for the first player). If it is true, then this method should be next to the steps that places the system into this expected state, otherwise it only overcomplicates things when you move it into a separate file.

File: E2E/tests/SPCounting.test.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+Scenario('Final SP is rounded to closest value when closest value is boundary value', async () => {
+  login.firstVoterLogin(username.user1);
+    let url = await I.grabCurrentUrl();
+    session(username.user2, () => {
+      login.voterLoginIntoCreatedGameRoom(url, username.user2);
+      game.vote(ChessPiece.queen, ChessTile.f5);
+    });
+    game.vote(ChessPiece.pawn, ChessTile.b3);
+    await assertions.checkIfTheFinalSPIsRoundedCorrectly();
```
Comment (jvalatkevicius) [inline E2E/tests/SPCounting.test.ts:29]: It is not clear what the test is expecting for, this should be written in some other way.

Comment (jvalatkevicius) [reply E2E/tests/SPCounting.test.ts:29]: I would expect to see something like this
```
assertions.finalSPAfterRoundingEqualsTo(5 (or whatever number is expected))
```

This makes the assertion reusable and the test more readable.

File: E2E/assertions/assertions.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  },
+
+  async checkIfTheSPIsRoundedCorrectlyForTheSecondPlayer() {
+    const averageSecondPlayerSP = game.calculateAverage("queen", "f", 5);
+    let expectedSecondPlayerScore = game.checkPlayerSP(averageSecondPlayerSP);
+    let actualSecondPlayerSP = Number(await I.grabTextFrom(game.locator.playersList.playerIndividualSP(1)));
+    I.assertEqual(actualSecondPlayerSP, expectedSecondPlayerScore);
+  },
+
+  async checkIfTheFinalSPIsRoundedCorrectly() {
```
Comment (jvalatkevicius) [inline E2E/assertions/assertions.ts:26]: Does it only allows to assert result if there are 2 players?

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: E2E/assertions/assertions.ts (hunk 6 above)
Comment (RutaB98) [reply E2E/assertions/assertions.ts:26]: Yes

File: E2E/assertions/assertions.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      I.usePlaywrightTo('check clipboard', async ({ page, browserContext }) => {
+        await browserContext.grantPermissions(['clipboard-read']);
+        const clipboardText = await page.evaluate("navigator.clipboard.readText()");
+        I.assertEqual(currentUrl, clipboardText);
+      });
+  },
+
+  async checkIfTheSPIsRoundedCorrectlyForTheFirstPlayer() {
+    const averageFirstPlayerSP = game.calculateAverage("bishop", "d", 3);
+    let expectedFirstPlayerScore = game.checkPlayerSP(averageFirstPlayerSP);
```
Comment (jvalatkevicius) [inline E2E/assertions/assertions.ts:14]: why let is used?

File: E2E/assertions/assertions.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    let actualFirstPlayerSP = Number(await I.grabTextFrom(game.locator.playersList.playerIndividualSP(2)));
```
Comment (jvalatkevicius) [inline E2E/assertions/assertions.ts:15]: `Number(await I.grabTextFrom(game.locator.playersList.playerIndividualSP(2))` - should be a separate function

File: E2E/assertions/assertions.ts (hunk 7 above)
Comment (jvalatkevicius) [inline E2E/assertions/assertions.ts:14]: Join these two into a single function
```
game.calculateAverage("bishop", "d", 3);
game.checkPlayerSP(averageFirstPlayerSP);
```

Comment (jvalatkevicius) [inline E2E/assertions/assertions.ts:14]: `checkPlayerSP` should have a better name, it tells nothing about what the function actually does

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: E2E/pageObjects/game.page.ts [hunk 9]
```diff
@@ -124,15 +128,50 @@ export = {
...
+    const averageSPList = [1.5, 2.5, 4, 6.5, 10.5];
+    let expectedSP = 1;
+    for(let i = 0; averageSPList[i] <= averageScore; i++){
+      expectedSP = sp[i + 1];
+    };
+    return expectedSP;
+  },
+
+  calculateAverage:(piece: string, letter: string, number: number) => {
+    let piecesScore = {pawn:1, knight:2, bishop:3, rook:5, king:8, queen:13};
```
Comment (KristinaKrev) [inline E2E/pageObjects/game.page.ts:153]: We have chess pieces names in test_data/ChessPieces.ts file, that can be reused, or maybe even numbers from ChessPieceValue?

**Review by KristinaKrev — DISMISSED** [review]

**Review by jvalatkevicius — DISMISSED** [review]

**Review by jvalatkevicius — APPROVED** [review]

**Review by AndzejIlj — APPROVED** [review]
